icann-rdap-client/src/gtld/entity.rs

```diff
@@ -0,0 +1,225 @@
+                                    }
+                                    if !role_info.org.is_empty() {
+                                        formatted_data += &format!(
+                                            "{} Organization: {}\n",
+                                            cfl(role),
+                                            role_info.org
+                                        );
+                                    }
+                                    if !role_info.adr.is_empty() {
+                                        formatted_data += &role_info.adr;
+                                    }
+                                }
+                            }
+                            _ => {} // Are there any roles we are missing?
+                        }
+                    }
+                }
+            }
+        }
+
+        front_formatted_data += &formatted_data;
+        front_formatted_data
+    }
+}
+
+fn format_address_with_label(
+    params: &mut GtldParams,
+    address_components: &[serde_json::Value],
+) -> String {
+    let postal_address = PostalAddress::builder()
+        .street_parts(
+            address_components
+                .get(2)
+                .and_then(|v| v.as_str())
+                .map_or_else(Vec::new, |s| vec![s.to_string()]),
+        )
+        .locality(
+            address_components
+                .get(3)
+                .and_then(|v| v.as_str())
+                .map_or_else(String::new, String::from),
+        )
+        .region_name(
+            address_components
+                .get(4)
+                .and_then(|v| v.as_str())
+                .map_or_else(String::new, String::from),
+        )
+        .country_name(
+            address_components
+                .get(6)
+                .and_then(|v| v.as_str())
+                .map_or_else(String::new, String::from),
+        )
+        .country_code(
+            address_components
+                .get(6)
+                .and_then(|v| v.as_str())
+                .map_or_else(String::new, String::from),
+        )
+        .postal_code(
+            address_components
+                .get(5)
+                .and_then(|v| v.as_str())
+                .map_or_else(String::new, String::from),
+        )
+        .build();
+
+    postal_address.to_gtld(params).to_string()
+}
+
+fn extract_role_info(
+    role: &str,
+    vcard_array: &[serde_json::Value],
+    params: &mut GtldParams,
+) -> RoleInfo {
+    let mut name = String::new();
+    let mut org = String::new();
+    let mut url = String::new();
+    let mut adr = String::new();
+
+    let label = match role {
+        "registrar" => "Registrar",
+        "technical" => "Technical",
+        "administrative" => "Admin",
+        "registrant" => "Registrant",
+        _ => "",
+    };
+    params.label = label.to_string();
+
+    for vcard in vcard_array.iter() {
```

Comment:
I think it would be better to extend the Contact struct to support any missing features than to directly use the vcard. That way switching to JSContact or SimpleContact will be more straight forward for this code if/when that occurs.